src/notebooksyncagent.cpp

```diff
@@ -43,29 +41,63 @@
 #include <journal.h>
 #include <attendee.h>
 
+#include <QUuid>
 #include <QDebug>
 
 
-#define NOTEBOOK_FUNCTION_CALL_TRACE FUNCTION_CALL_TRACE(QString("%1 %2").arg(Q_FUNC_INFO).arg(mNotebook ? mNotebook->account() : ""))
+#define NOTEBOOK_FUNCTION_CALL_TRACE FUNCTION_CALL_TRACE(QString("%1 %2").arg(Q_FUNC_INFO).arg(m_notebook ? m_notebook->account() : ""))
+
+namespace {
+    // KCalCore API only got support for remote url in version 4.12
+    // Our version of KCalCore is too old and doesn't support that.
+    // Instead, we use custom properties - but these get purged when the
+    // incidence is deleted, so we can't know what the uri is for a deleted
+    // incidence except by using some trickery (see calculateDelta()).
+    // In the future, if we upgrade, change the following to incidence->url().
+    QString incidenceHrefUri(KCalCore::Incidence::Ptr incidence, const QString &remoteCalendarPath = QString(), bool *uriNeedsFilling = 0)
+    {
+        QString remoteUri = incidence->customProperty("buteo", "uri");
+        // if the uriNeedsFilling param is supplied, we build what we think the uri "should" be.
+        if (remoteUri.isEmpty() && uriNeedsFilling && !remoteCalendarPath.isEmpty()) {
+            *uriNeedsFilling = true;
+            remoteUri = remoteCalendarPath + incidence->uid() + ".ics";
```

Comment:
Hrm.  You are correct - if two separate devices are involved, and one of them uploads the file to something other than remotepath/UID.ics then there is indeed a problem.

How should we resolve this?  Should I store URI into the COMMENTS instead of the custom property, as I did in the Google PR?  Should I do that for ETAG also?